// random/dist/generators/xor128.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var rng_1 = __importDefault(require("../rng"));
var RNGXOR128 = /** @class */ (function (_super) {
    __extends(RNGXOR128, _super);
    function RNGXOR128(seed, opts) {
        var _this = _super.call(this) || this;
        _this.x = 0;
        _this.y = 0;
        _this.z = 0;
        _this.w = 0;
        _this.seed(seed, opts);
        return _this;
    }
    Object.defineProperty(RNGXOR128.prototype, "name", {
        get: function () {
            return 'xor128';
        },
        enumerable: false,
        configurable: true
    });
    RNGXOR128.prototype.next = function () {
        var t = this.x ^ (this.x << 1);
        this.x = this.y;
        this.y = this.z;
        this.z = this.w;
        this.w = this.w ^ ((this.w >>> 19) ^ t ^ (t >>> 8));
        return (this.w >>> 0) / 0x100000000;
    };
    RNGXOR128.prototype.seed = function (seed, opts) {
        this.x = this._seed(seed, opts);
        // discard an initial batch of 64 values
        for (var i = 0; i < 64; ++i) {
            this.next();
        }
    };
    RNGXOR128.prototype.clone = function (seed, opts) {
        return new RNGXOR128(seed, opts);
    };
    return RNGXOR128;
}(rng_1.default));
exports.default = RNGXOR128;
